src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs

```diff
@@ -274,19 +278,60 @@ internal NotRecognizedSemaphore(int initialCount, JoinableTaskContext joinableTa
             {
                 Requires.NotNull(operation, nameof(operation));
 
+                // Note: this code is duplicated and not extracted to minimize allocating extra async state machines.
+                // For performance reasons in the JTF enabled scenario, we want to minimize the number of Joins performed, and also
+                // keep the size of the JoinableCollection to a minimum. This also means awaiting on the semaphore outside of a
+                // JTF.RunAsync. This requires us to not ConfigureAwait(true) on the semaphore. However, that prevents us from
+                // resuming on the correct sync context. To partially fix this, we will at least resume you on the main thread or
+                // thread pool.
+                AsyncSemaphore.Releaser releaser;
+                bool usingJtf = this.joinableTaskFactory != null;
+                bool resumeOnMainThread = usingJtf ? this.joinableTaskCollection.Context.IsOnMainThread : false;
+                bool mustYield = false;
```

Comment:
These are identical in each concrete class and feel like they could be properties on the base class.  Is there any reason you didn't implement them that way?

Reply:
They are all locals. I could move `usingJtf` to be a readonly bool though.

Reply:
(I was commenting in CodeFlow and didn't realize that the highlighted lines don't seem to show up very well.)

I was referring to ```usingJtf``` and ```resumeOnMainThread```.  Those definitions are identical in all four concrete types.  You could make ```usingJtf``` a ```readonly bool```, but I'm not sure it's worth the extra field.  I was envisioning something like this in ```ReentrantSemaphore```:

```cs
protected bool UsingJtf => this.joinableTaskFactory != null;
protected bool ResumeOnMainThread => this.UsingJtf ? this.joinableTaskCollection.Context.IsOnMainThread : false;
```

Not a huge deal, though.

Reply:
I will extract `UsingJtf` to a property like you suggested, that is a good idea. `ResumeOnMainThread` is not much of a benefit as it must captured at that point in time. I think capturing a property at that point will be less descriptive of what I am trying to achieve there.